### PR TITLE
Improve thread scheduling for lockhammer

### LINKED ISSUE #16: Setting affinity before creating threads
How about setting the affinity before creating the threads.
I see that 
a. tests are running faster
b. with many cores > 200, tests use to hang with thread0 stuck in wait.
by setting affinity before creating threads, no hang is observed

### LINKED ISSUE #17: Add a core-number interleaving command line option
Often on systems with multiple hardware threads the logical core numbering is assigned such that lower numbered cores reference the first hardware thread on each core and higher numbered cores reference subsequent hardware threads on those same cores.  For example, on a 4-core 8-thread system logical core 0 references the first thread on physical core 0, logical core 1 references  the first thread on physical core 1, logical core 4 references the second thread on physical core 0, and so on.  Currently lockhammer adjusts its core population to fill up all threads on each physical cores first since these but does so in a way that only works correctly for 2 thread-per-core systems.  A generic mechanism for specifying the number of "regions" of logical core numbering should be added and a way to specify the correct value on the command line provided.

### PR DESCRIPTION
**Pull request by lucasclucasdo**
Add a core interleave argument to lockhammer which determines the order in which logical core numbers are populated.

Fixes #17

### CODE REVIEW

**Comment by geoffreyblake** [comment]
Can one of the admins verify this patch?

**Comment by lucasclucasdo** [comment]
Let's rethink this one since it didn't fix the issue Cavium raised.  The interleave option is still useful for other reasons but I'd rather not change the affinity setting unless we have a good reason to do so.

**Comment by lucasclucasdo** [comment]
I have updated this PR to only include the interleave command line option.  We can address further thread scheduling issues in subsequent PRs after further experimentation.

File: benchmarks/lockhammer/src/lockhammer.c [hunk 1]
```diff
@@ -265,7 +278,7 @@ void* hmr(void *ptr)
     }
     else {
         /* Calculate affinity mask for my core and set affinity */
-        CPU_SET(((mycore >> 1)) + ((ncores >> 1) * (mycore & 1)), &affin_mask);
+        CPU_SET(((mycore * ncores / ileave) % ncores + (mycore / ileave)), &affin_mask);
```
Comment (geoffreyblake) [inline benchmarks/lockhammer/src/lockhammer.c]: Hi Lucas,
Can you add a comment here about the interleaving assumption is 0 - (numcores / numthreads) is thread 0 on each physical core, (numcores / numthreads) - 2 * (numcores / numthreads) is thread 1 on each core etc?

**Pull request merged**